### PR TITLE
Bay Porting Phase 2: Movement.   Part 1: Move handlers

### PR DESCRIPTION
**Pull request by NanakoAC**
This PR is the first part of Phase 2 of bay porting, the movement system.

This project cuts out more or less our entire movement system and replaces it with bay's latest version, whole cloth.

Most of this PR is internal nerd work that players won't care about but there are a couple user facing changes. Most notable is the addition of crawling. Players can now drag themselves along the ground fairly slowly while lying down, assuming they're still conscious

Move speeds may have changed a bit in a variety of mobs and animals. I tested most things and found the speeds to be fairly sane, but not everything is identical. Normal running speed should be about the same, walking is faster, other things may vary. A lot of testing is needed

I've spent the past 3-4 days mostly just testing this PR, and i'm about 85% confident that most of it is probably in a pretty decent and useable condition. Probably nothing is broken, maybe

QA Personnel should test the following:

- AI controlled rigs
- shuttles
- Animal/critter speeds

------------------------------------------------------------------------------------------------------------------
Boring nerd stuff starts here

The main feature of this PR is bay's new move handler system. Instead of the horrible old method of an ever growing monstrosity of code in client/Move, now every mob gets their own list of movement handlers in various categories. Before each step, each of these handlers is asked for permission to make the step, whereby it does its appropriate checks (conscious, time since last move, etc)

By my reckoning, this removes approximately 35-50% of the awful inefficient mess that is movement. It is a huge step forward and is certainly better.

Other small changes: 
Silenced var removed on tools and shoes, now it uses an item flag instead. Bitfields are faster and this is important for movement code

The movement system is far from perfect though, a lot of awfulness remains. I wouldn't even call it "good" from an objective viewpoint, it is just "less terrible". But it does provide a good foundation to make a near perfect system with some more work, and so this PR will not be the last on this subject. Several major flaws still exist that bay has not fixed, including:

- Move delay is still recalculated with every step
- Many of the move handlers don't need to exist at all times, and could be intelligently added or removed only under appropriate circumstances
- Move handlers are constantly checked and if necessary, initialised with every step
- Movement speed is still expressed in a delay of time, and not in a speed

Once this PR is all tested and merged, I intend to make at least one, possibly two more, addressing the above issues, and also bringing a few more interesting features for users, namely:
Sprinting
A proper encumbrance system for worn equipment

### CODE REVIEW

File: code/__defines/error_handler.dm [hunk 1]
```diff
@@ -2,6 +2,6 @@
 #define ERROR_COOLDOWN 600 // The "cooldown" time for each occurrence of a unique error
 #define ERROR_LIMIT 9 // How many occurrences before the next will silence them
 #define ERROR_MAX_COOLDOWN (ERROR_COOLDOWN * ERROR_LIMIT)
-#define ERROR_SILENCE_TIME 6000 // How long a unique error will be silenced for
+#define ERROR_SILENCE_TIME 6000 // How long a unique error will be item_flags & SILENT for
```
Comment (NanakoAC) [inline code/__defines/error_handler.dm]: oops, i mass replaced "silenced" with this bitfield check

File: code/__defines/error_handler.dm [hunk 2]
```diff
@@ -2,6 +2,6 @@
 #define ERROR_COOLDOWN 600 // The "cooldown" time for each occurrence of a unique error
 #define ERROR_LIMIT 9 // How many occurrences before the next will silence them
 #define ERROR_MAX_COOLDOWN (ERROR_COOLDOWN * ERROR_LIMIT)
-#define ERROR_SILENCE_TIME 6000 // How long a unique error will be silenced for
+#define ERROR_SILENCE_TIME 6000 // How long a unique error will be silences for
```
Comment (mattroks101) [inline code/__defines/error_handler.dm]: Mass replace messed up the comment.

File: code/game/objects/items/weapons/tools/_tools.dm [hunk 3]
```diff
@@ -682,12 +681,13 @@
...
 	switched_on_force = initial(switched_on_force)
 	extra_bulk = initial(extra_bulk)
-	silenced = initial(silenced)
+	item_flags = initial(item_flags)
 	name = initial(name)
 	max_upgrades = initial(max_upgrades)
 	color = initial(color)
 	sharp = initial(sharp)
 	prefixes = list()
+	item_flags = initial(item_flags)
```
Comment (mattroks101) [inline code/game/objects/items/weapons/tools/_tools.dm]: You set item flags twice here.

File: code/game/objects/structures/stool_bed_chair_nest/wheelchair.dm [hunk 4]
```diff
@@ -103,11 +103,11 @@
 					unbuckle_mob()
 			if (pulling && (get_dist(src, pulling) > 1))
 				pulling.pulledby = null
-				pulling << SPAN_WARNING("You lost your grip!")
+				to_chat(pulling, "<span class='warning'>You lost your grip!</span>")
```
Comment (mattroks101) [inline code/game/objects/structures/stool_bed_chair_nest/wheelchair.dm]: Lost a SPAN macro here.

File: code/modules/mob/living/carbon/human/human_movement.dm [hunk 5]
```diff
@@ -1,67 +1,44 @@
...
+	if (istype(loc, /turf/space)) // It's hard to be slowed down in space by... anything
+		return tally
 
 	if(embedded_flag)
 		handle_embedded_objects() //Moving with objects stuck in you can cause bad times.
-
 	if(CE_SPEEDBOOST in chem_effects)
-		return 0
+		tally -= chem_effects[CE_SPEEDBOOST]
+	//if(CE_SLOWDOWN in chem_effects)
```
Comment (mattroks101) [inline code/modules/mob/living/carbon/human/human_movement.dm]: Any reason this is commented out and not deleted?

File: code/modules/mob/mob_movement.dm [hunk 6]
```diff
@@ -46,33 +59,36 @@
...
 
 	if(!usr.pulling)
-		usr << "\blue You are not pulling anything."
+		to_chat(usr, "<span class='notice'>You are not pulling anything.</span>")
 		return
 	usr.stop_pulling()
 
 /client/verb/swap_hand()
 	set hidden = 1
-	if(iscarbon(mob))
```
Comment (mattroks101) [inline code/modules/mob/mob_movement.dm]: Don't remove our `iscarbon()` proc please.

File: code/modules/mob/mob_movement.dm [hunk 7]
```diff
@@ -46,33 +59,36 @@
...
+	if(istype(mob, /mob/living/carbon))
 		mob:swap_hand()
-	if(isrobot(mob))
```
Comment (mattroks101) [inline code/modules/mob/mob_movement.dm]: Same with our `isrobot()`

File: code/modules/mob/mob_movement_old.dm [hunk 8]
```diff
@@ -0,0 +1,529 @@
+/mob/CanPass(atom/movable/mover, turf/target, height=0, air_group=0)
```
Comment (mattroks101) [inline code/modules/mob/mob_movement_old.dm]: Is this whole file actually used for anything?

Comment (NanakoAC) [reply code/modules/mob/mob_movement_old.dm]: it is not anymore, i just had it as a reference for integrating the new system to existing code, i will remove it

**Comment by NanakoAC** [comment]
Fixed requested changes, ready for round 2, and also testing

File: code/game/objects/effects/overlays.dm [hunk 9]
```diff
@@ -3,6 +3,7 @@
 	unacidable = 1
 	var/i_attached//Added for possible image attachments to objects. For hallucinations and the like.
 
+
```
Comment (mattroks101) [inline code/game/objects/effects/overlays.dm]: This file doesn't appear to involve the bay movement stuff.

File: code/game/objects/structures/stool_bed_chair_nest/wheelchair.dm [hunk 10]
```diff
@@ -31,7 +34,7 @@
 		if(user==pulling)
 			pulling = null
 			user.pulledby = null
-			user << SPAN_WARNING("You lost your grip!")
+			to_chat(user, "<span class='warning'>You lost your grip!</span>")
```
Comment (mattroks101) [inline code/game/objects/structures/stool_bed_chair_nest/wheelchair.dm]: I don't think deleted the span macros was intentional.

File: code/game/objects/structures/stool_bed_chair_nest/wheelchair.dm [hunk 11]
```diff
@@ -120,17 +120,17 @@
 	if(in_range(src, user))
 		if(!ishuman(user))	return
 		if(user == buckled_mob)
-			user << SPAN_WARNING("You realize you are unable to push the wheelchair you sit in.")
+			to_chat(user, "<span class='warning'>You realize you are unable to push the wheelchair you sit in.</span>")
```
Comment (mattroks101) [inline code/game/objects/structures/stool_bed_chair_nest/wheelchair.dm]: Lost it here too.

File: code/game/objects/structures/stool_bed_chair_nest/wheelchair.dm [hunk 12]
```diff
@@ -145,32 +145,29 @@
...
 		var/blocked = occupant.run_armor_check(def_zone, "melee")
-		occupant.throw_at(A, 3, propelled)
+		occupant.throw_at(A, 3, 3, propelled)
 		occupant.apply_effect(6, STUN, blocked)
 		occupant.apply_effect(6, WEAKEN, blocked)
 		occupant.apply_effect(6, STUTTER, blocked)
-		occupant.apply_damage(10, BRUTE, def_zone)
+		occupant.apply_damage(10, BRUTE, def_zone, blocked)
 		playsound(src.loc, 'sound/weapons/punch1.ogg', 50, 1, -1)
-		if(isliving(A))
```
Comment (mattroks101) [inline code/game/objects/structures/stool_bed_chair_nest/wheelchair.dm]: Lost an `isliving()`

File: code/game/objects/structures/stool_bed_chair_nest/wheelchair.dm [hunk 13]
```diff
@@ -145,32 +145,29 @@
...
-			occupant.visible_message(SPAN_DANGER("[pulling] has thrusted \the [name] into \the [A], throwing \the [occupant] out of it!"))
-
-			pulling.attack_log += "\[[time_stamp()]\]<font color='red'> Crashed [occupant.name]'s ([occupant.ckey]) [name] into \a [A]</font>"
-			occupant.attack_log += "\[[time_stamp()]\]<font color='orange'> Thrusted into \a [A] by [pulling.name] ([pulling.ckey]) with \the [name]</font>"
-			msg_admin_attack("[pulling.name] ([pulling.ckey]) has thrusted [occupant.name]'s ([occupant.ckey]) [name] into \a [A] (<A HREF='?_src_=holder;adminplayerobservecoodjump=1;X=[pulling.x];Y=[pulling.y];Z=[pulling.z]'>JMP</a>)")
+			occupant.visible_message("<span class='danger'>[pulling] has thrusted \the [name] into \the [A], throwing \the [occupant] out of it!</span>")
+			admin_attack_log(pulling, occupant, "Crashed their victim into \an [A].", "Was crashed into \an [A].", "smashed into \the [A] using")
 		else
-			occupant.visible_message(SPAN_DANGER("[occupant] crashed into \the [A]!"))
+			occupant.visible_message("<span class='danger'>[occupant] crashed into \the [A]!</span>")
```
Comment (mattroks101) [inline code/game/objects/structures/stool_bed_chair_nest/wheelchair.dm]: And another SPAN macro.

File: code/modules/mob/mob_movement.dm [hunk 14]
```diff
@@ -46,35 +59,39 @@
 				var/mob/living/carbon/C = usr
 				C.toggle_throw_mode()
 			else
-				usr << "\red This mob type cannot throw items."
+				to_chat(usr, "<span class='warning'>This mob type cannot throw items.</span>")
```
Comment (mattroks101) [inline code/modules/mob/mob_movement.dm]: Might wanna use a SPAN macro here since we have them.

File: code/modules/mob/mob_movement.dm [hunk 15]
```diff
@@ -46,35 +59,39 @@
...
+	else
+		unEquip(get_active_hand(), loc)
 
 //This gets called when you press the delete button.
 /client/verb/delete_key_pressed()
 	set hidden = 1
 
 	if(!usr.pulling)
-		usr << "\blue You are not pulling anything."
+		to_chat(usr, "<span class='notice'>You are not pulling anything.</span>")
```
Comment (mattroks101) [inline code/modules/mob/mob_movement.dm]: Same here.

File: code/modules/mob/mob_movement.dm [hunk 16]
```diff
@@ -88,7 +105,7 @@
 
 /client/verb/toggle_throw_mode()
 	set hidden = 1
-	if(!iscarbon(mob))
+	if(!istype(mob, /mob/living/carbon))
```
Comment (mattroks101) [inline code/modules/mob/mob_movement.dm]: Lost the `iscarbon()`

File: code/modules/mob/mob_movement.dm [hunk 17]
```diff
@@ -103,341 +120,82 @@
...
-	if(restrained()) //Check to see if we can do things
-		return 0
 
-	//Check to see if we slipped
-	if(prob(slip_chance(5)) && !buckled)
-		src << SPAN_WARNING("You slipped!")
+//return 1 if slipped, 0 otherwise
+/mob/proc/handle_spaceslipping()
+	if(prob(5)) //Todo: Factor in future agility stat here
+		to_chat(src, "<span class='warning'>You slipped!</span>")
```
Comment (mattroks101) [inline code/modules/mob/mob_movement.dm]: SPAN macro could go here.

**Comment by NanakoAC** [comment]
Another round of requested changes done, and i see there's a conflict too, ill get that

i'm also going to retarget this at development since it'll need some testing before going live

File: code/game/objects/structures/stool_bed_chair_nest/wheelchair.dm [hunk 18]
```diff
@@ -49,10 +52,10 @@
 		if(user==pulling)
 			return
 	if(pulling && (get_dir(src.loc, pulling.loc) == direction))
-		user << SPAN_WARNING("You cannot go there.")
+		to_chat(user, "<span class='warning'>You cannot go there.</span>")
```
Comment (mattroks101) [inline code/game/objects/structures/stool_bed_chair_nest/wheelchair.dm]: Still lost a span macro here

File: code/game/objects/structures/stool_bed_chair_nest/wheelchair.dm [hunk 19]
```diff
@@ -49,10 +52,10 @@
...
 		return
 	if(pulling && buckled_mob && (buckled_mob == user))
-		user << SPAN_WARNING("You cannot drive while being pushed.")
+		to_chat(user, "<span class='warning'>You cannot drive while being pushed.</span>")
```
Comment (mattroks101) [inline code/game/objects/structures/stool_bed_chair_nest/wheelchair.dm]: And here.

**Comment by NanakoAC** [comment]
requested changes done and fixed

**Comment by mattroks101** [comment]
Is it all good then?

**Comment by NanakoAC** [comment]
This has been on the test server and tested by a team of people, i've fixed every problem they found and I now have 100% confidence that we can skip development, i'll retarget again to master branch

**Comment by NanakoAC** [comment]
This project is basically done aside from retargeting, i'll leave it as it is for someone else to make the decision

**Comment by ACCount12** [comment]
I'm merging this into the development branch for now. This would make the eventual merge of dev branch into master even more painful, but it looks like it's going to be a pain anyways.

**Pull request merged**